src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/TimeSpanConverter.cs

```diff
@@ -0,0 +1,92 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Buffers.Text;
+using System.Diagnostics;
+
+namespace System.Text.Json.Serialization.Converters
+{
+    internal sealed class TimeSpanConverter : JsonConverter<TimeSpan>
+    {
+        private const int MinimumTimeSpanFormatLength = 8; // hh:mm:ss
+        private const int MaximumTimeSpanFormatLength = 26; // -dddddddd.hh:mm:ss.fffffff
+        private const int MaximumEscapedTimeSpanFormatLength = JsonConstants.MaxExpansionFactorWhileEscaping * MaximumTimeSpanFormatLength;
+
+        public override TimeSpan Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        {
+            if (reader.TokenType != JsonTokenType.String)
+            {
+                throw ThrowHelper.GetInvalidOperationException_ExpectedString(reader.TokenType);
+            }
+
+            bool isEscaped = reader._stringHasEscaping;
+            int maximumLength = isEscaped ? MaximumEscapedTimeSpanFormatLength : MaximumTimeSpanFormatLength;
+
+            ReadOnlySpan<byte> source = stackalloc byte[0];
+
+            if (reader.HasValueSequence)
+            {
+                ReadOnlySequence<byte> valueSequence = reader.ValueSequence;
+                long sequenceLength = valueSequence.Length;
+
+                if (!JsonHelpers.IsInRangeInclusive(sequenceLength, MinimumTimeSpanFormatLength, maximumLength))
+                {
+                    throw ThrowHelper.GetFormatException();
+                }
+
+                Span<byte> stackSpan = stackalloc byte[(int)sequenceLength];
```

Comment:
`sequenceLength` is bounded by `26 x 6 = 156`. Given the security risks associated with stack allocation in deserialization code, maybe add a comment documenting this invariant.

Perhaps though we might want to use the upper bound constants directly when stack allocating the buffer:
```csharp
Span<byte> stackSpan = stackallock byte[isEscaped ? MaximumEscapedTimeSpanFormatLength : MaximumTimeSpanFormatLength];
valueSequence.CopyTo(stackSpan);
source = stackSpan.Slice(0, sequenceLength);
```
Tagging @GrabYourPitchforks, who might have stronger opinions on what should be the preferred approach.

Reply:
Constant values are definitely preferred since they can be optimized by the JIT. If the JIT sees a variable length being passed into stackalloc, it doesn't have intrinsic knowledge what bounds that length might have, so it emits stack-probing code in a loop. There's no need for us to pay that cost.

Reply:
Keep in mind also that `Utf8Parser.TryParse(..., out TimeSpan, ...)` allows leading whitespace. In theory you could give it an arbitrarily long string and it would happily skip leading space and tab chars. Do you intend to support somebody passing `"    (lots o' spaces)    dddddddd.hh:mm:ss.fffffff"` as input?

Reply:
> Perhaps though we might want to use the upper bound constants directly when stack allocating the buffer:

Per @GrabYourPitchforks' note, let's take the approach suggested by @eiriktsarpalis:

```cs
Span<byte> stackSpan = stackallock byte[isEscaped ? MaximumEscapedTimeSpanFormatLength : MaximumTimeSpanFormatLength];
valueSequence.CopyTo(stackSpan);
source = stackSpan.Slice(0, sequenceLength);
```

There are other places in the JSON codebase that can benefit from this suggestion, e.g. [here](https://github.com/dotnet/runtime/blob/09785ef78f8c758350ec1ded604484bbbed0ce99/src/libraries/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.TryGet.cs#L1087). We can update them in this PR or in a follow up.

> Do you intend to support somebody passing " (lots o' spaces) dddddddd.hh:mm:ss.fffffff" as input?

To be consistent with STJ's `DateTime(Offset)` parsing, let's disallow and add tests for leading/trailing whitespace. I believe the only change needed is to verify that the first token in the unescaped buffer is a number or hypen (`-`).

Reply:
Updated